src/PairedUrlStructure.php

```diff
@@ -0,0 +1,59 @@
+<?php
+/**
+ * Abstract class PairedUrl.
+ *
+ * @package AmpProject\AmpWP
+ */
+
+namespace AmpProject\AmpWP;
+
+/**
+ * Interface for classes that implement a PairedUrl.
+ *
+ * @package AmpProject\AmpWP
+ * @since 2.1
+ */
+abstract class PairedUrlStructure {
+
+	/**
+	 * Paired URL service.
+	 *
+	 * @var PairedUrl
+	 */
+	protected $paired_url;
+
+	/**
+	 * PairedUrlStructure constructor.
+	 *
+	 * @param PairedUrl $paired_url Paired URL service.
+	 */
+	public function __construct( PairedUrl $paired_url ) {
+		$this->paired_url = $paired_url;
+	}
+
+	/**
+	 * Determine a given URL is for a paired AMP request.
+	 *
+	 * @param string $url URL (or REQUEST_URI).
+	 * @return bool True if the URL has the paired endpoint.
+	 */
+	public function has_endpoint( $url ) {
+		return $url !== $this->remove_endpoint( $url );
+	}
```

Comment:
This logic is never used as it is overridden in all 4 implementations. We should make this abstract as well.
```suggestion
	abstract public function has_endpoint( $url );
```

Reply:
See examples in custom paired URL structures:

* https://gist.github.com/westonruter/67487d199f4cbc1968a30c60de121893#file-subdomainurlstructure-php
* https://gist.github.com/westonruter/f881658cb1bc5a03cc432c8e2b89102a#file-prefixurlstructure-php

I haven't removed the `has_endoint` method from the bundled paired URL structures yet since I wasn't sure yet if there's some value in having tailored `has_endpoint` implementations.